### ISSUE TITLE
bug: formatter running when there is no config for the formatter in the project I'm in

### ISSUE THREAD

**Issue by NullVoxPopuli** [stevearc/conform.nvim#268]
### Neovim version (nvim -v)

NVIM v0.10.0-dev-2042+g28c183b38

### Operating system/version

Ubuntu 23.10

### Add the debug logs

- [ ] I have set `log_level = vim.log.levels.DEBUG` and pasted the log contents below.

### Log file

I don't know how to set log level

### Describe the bug

My repo doesn't have a prettier config, yet prettier is still running.

### What is the severity of this bug?

tolerable (can work around it)

### Steps To Reproduce

My config:
```lua
  use {
    'stevearc/conform.nvim',
    config = function()
      require('conform').setup({
        formatters_by_ft = {
          -- Use a sub-list to run only the first available formatter
          javascript = { { "prettierd", "prettier" } },
          ['javascript.glimmer'] = { { "prettierd", "prettier" } },
        },
        format_on_save = {
          timeout_ms = 250,
          lsp_fallback = true,
        },
        notify_on_error = false,
      })
    end
```

### Expected Behavior

When there is no prettier config in the repo, prettier should not run

### Minimal example file

- clone https://github.com/starbeamjs/dev-compile
- have the above conform config
- have prettierd installed
- save any file, see that changes occur

### Minimal init.lua

```Lua
-- DO NOT change the paths and don't remove the colorscheme
local root = vim.fn.fnamemodify("./.repro", ":p")

-- set stdpaths to use .repro
for _, name in ipairs({ "config", "data", "state", "cache" }) do
  vim.env[("XDG_%s_HOME"):format(name:upper())] = root .. "/" .. name
end

-- bootstrap lazy
local lazypath = root .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({
    "git",
    "clone",
    "--filter=blob:none",
    "--single-branch",
    "https://github.com/folke/lazy.nvim.git",
    lazypath,
  })
end
vim.opt.runtimepath:prepend(lazypath)

-- install plugins
local plugins = {
  "folke/tokyonight.nvim",
  {
    "stevearc/conform.nvim",
    config = function()
      require("conform").setup({
        log_level = vim.log.levels.DEBUG,
        -- add your config here
      })
    end,
  },
  -- add any other plugins here
}
require("lazy").setup(plugins, {
  root = root .. "/plugins",
})

vim.cmd.colorscheme("tokyonight")
-- add anything else here
```

**Comment by stevearc** [comment]
Prettier by default will look for these files to determine the root directory, and run in that directory:
https://github.com/stevearc/conform.nvim/blob/0e61fc88f725b4411e63b43470a397c6102534c7/lua/conform/formatters/prettier.lua#L63-L76

If you want to _only_ run when one of those files is detected, you can modify the formatter to set `require_cwd = true`.

```lua
require("conform").setup({
  formatters = {
    prettier = {
      require_cwd = true
    }
  }
})
```

**Comment by NullVoxPopuli** [comment]
seems like this is the problem: https://github.com/stevearc/conform.nvim/blob/master/lua/conform/formatters/prettierd.lua#L26
presence of package.json is not sufficient to enable prettier :thinking:

**Comment by NullVoxPopuli** [comment]
@stevearc is it possible to change what files are used? I've set this up for my config, but I'm seeing that "some default prettier" config is used in formatjs:
```lua
  use {
    'stevearc/conform.nvim',
    config = function()
      local util = require("conform.util")

      require('conform').setup({
        -- We don't want to enable prettierd
        -- because it requires global installation, and then
        -- we can't even have projects without prettier
        --
        -- NOTE: make sure prettier (and prettierd) are not installed globally
        formatters_by_ft = {
          -- Use a sub-list to run only the first available formatter
          javascript = { { "prettier" } },
          ['javascript.glimmer'] = { { "prettier" } },
        },
        format_on_save = {
          timeout_ms = 250,
          lsp_fallback = true,
        },
        notify_on_error = false,
        formatters = {
          prettier = {
            require_cwd = true,
            cwd = util.root_file({
              -- https://prettier.io/docs/en/configuration.html
              ".prettierrc",
              ".prettierrc.json",
              ".prettierrc.yml",
              ".prettierrc.yaml",
              ".prettierrc.json5",
              ".prettierrc.js",
              ".prettierrc.cjs",
              ".prettierrc.mjs",
              ".prettierrc.toml",
              "prettier.config.js",
              "prettier.config.cjs",
              "prettier.config.mjs",
              -- "package.json",
            }),
          }
        }
      })
    end
  }
```

The repo I'm working in atm: https://github.com/formatjs/formatjs/

There is a repo-root prettier config, but it seems to be stopping at the nearest package.json and choosing default options, which isn't correct :thinking:

**Comment by gegoune** [comment]
@stevearc Apologies for piggybacking here but I am wondering if you considered having more informed check. One that would see if tool is configured within project configuration files such as `package.json` or `pyproject.toml`. Not all configurations can easily be determined simply by checking for existence of specific configuration files (although that's very healthy default I think).

This question also touches on https://github.com/stevearc/conform.nvim/issues/367, as it seems pretty wasteful to check content of configuration files on each format (I am not sure if conform.nvim sets formatters on buffer load or invocation of format itself). I'll see if I can come up with some solution that caches results.

**Comment by csntr** [comment]
@NullVoxPopuli I have the same issue as well, I got to this point with around the same config as you based on #407 . I don't know yet how I can debug formatting and see what exactly happened when I did `:w`.

**Comment by ic-768** [comment]
Same issue here

**Comment by pmanole-volvocars** [comment]
Had the same issue and I fixed it with the changes I have proposed in #567

**Comment by wellavelino** [comment]
interesting same issue, conform will actually respect prettier config, if being executed from a shortcut, however even after disabling `format_on_save` conform still runs the `default` configuration and discards the formatted one.
